File: pidex/sources/ssh.py
import re
import time
from datetime import datetime

from pidex.core.constants import EVENT_SSH_BRUTEFORCE, EVENT_SSH_LOGIN, EVENT_SSH_LOGOUT, SEVERITY_INFO, SEVERITY_WARN, SOURCE_SSH
from pidex.core.event import Event

_ACCEPTED_RE = re.compile(
    r"Accepted (?:publickey|password|keyboard-interactive) for (\S+) from (\S+)"
)
_FAILED_RE = re.compile(
    r"Failed (?:password|publickey) for (\S+) from (\S+)"
)
_DISCONNECTED_RE = re.compile(
    r"Disconnected from (?:user\s+)?(\S+) (\S+)"
)

_BRUTEFORCE_WINDOW = 30
_BRUTEFORCE_THRESHOLD = 3


def parse(entry: dict) -> Event | None:
    if entry.get("_COMM") != "sshd" and entry.get("SYSLOG_IDENTIFIER") != "sshd":
        return None

    message = entry.get("MESSAGE", "")

    m = _ACCEPTED_RE.search(message)
    if m:
        user, ip = m.group(1), m.group(2)
        return Event(
            source=SOURCE_SSH,
            event_type=EVENT_SSH_LOGIN,
            severity=SEVERITY_INFO,
            title="SSH Login",
            message=f"{user} logged in from {ip}",
            timestamp=_ts(entry),
        )

    m = _DISCONNECTED_RE.search(message)
    if m:
        user, ip = m.group(1), m.group(2)
        return Event(
            source=SOURCE_SSH,
            event_type=EVENT_SSH_LOGOUT,
            severity=SEVERITY_INFO,
            title="SSH Logout",
            message=f"{user} disconnected from {ip}",
            timestamp=_ts(entry),
        )

    m = _FAILED_RE.search(message)
    if m:
        user, ip = m.group(1), m.group(2)
        if _is_bruteforce(ip, entry):
            return Event(
                source=SOURCE_SSH,
                event_type=EVENT_SSH_BRUTEFORCE,
                severity=SEVERITY_WARN,
                title="SSH Brute Force",
                message=f"Repeated failed attempts for {user} from {ip}",
                timestamp=_ts(entry),
            )

    return None


_bruteforce_tracker: dict[str, list[float]] = {}


def _is_bruteforce(ip: str, entry: dict) -> bool:
    now = time.time()
    attempts = _bruteforce_tracker.setdefault(ip, [])
    attempts.append(now)
    cutoff = now - _BRUTEFORCE_WINDOW
    _bruteforce_tracker[ip] = [t for t in attempts if t > cutoff]
    return len(_bruteforce_tracker[ip]) >= _BRUTEFORCE_THRESHOLD


def _ts(entry: dict) -> datetime:
    micros = entry.get("__REALTIME_TIMESTAMP")
    if micros is not None:
        return datetime.fromtimestamp(int(micros) / 1_000_000)
    return datetime.now()

File: internal/parser/ssh.go
package parser

import (
	"regexp"
	"sync"
	"time"

	"github.com/Shivamingale3/pi_dex/internal/core"
)

var (
	acceptedRE = regexp.MustCompile(
		`Accepted (?:publickey|password|keyboard-interactive) for (\S+) from (\S+)`,
	)
	failedRE = regexp.MustCompile(
		`Failed (?:password|publickey) for (\S+) from (\S+)`,
	)
	disconnectedRE = regexp.MustCompile(
		`Disconnected from (?:user\s+)?(\S+) (\S+)`,
	)
)

const (
	bruteforceWindow    = 30
	bruteforceThreshold = 3
	bruteforceMaxIPs    = 1000
)

var bruteforceTracker struct {
	mu   sync.Mutex
	ips  map[string][]time.Time
	keys []string
}

func init() {
	bruteforceTracker.ips = make(map[string][]time.Time)
}

func ParseSSH(entry map[string]any) *core.Event {
	comm, _ := entry["_COMM"].(string)
	ident, _ := entry["SYSLOG_IDENTIFIER"].(string)
	if comm != "sshd" && comm != "sshd-session" && ident != "sshd" && ident != "sshd-session" {
		return nil
	}

	msg := entryMessage(entry)
	if msg == "" {
		return nil
	}

	if m := acceptedRE.FindStringSubmatch(msg); m != nil {
		return &core.Event{
			Source:    core.SourceSSH,
			EventType: core.EventSSHLogin,
			Severity:  core.SeverityInfo,
			Title:     "SSH Login",
			Message:   m[1] + " logged in from " + m[2],
			Timestamp: entryTimestamp(entry),
		}
	}

	if m := disconnectedRE.FindStringSubmatch(msg); m != nil {
		return &core.Event{
			Source:    core.SourceSSH,
			EventType: core.EventSSHLogout,
			Severity:  core.SeverityInfo,
			Title:     "SSH Logout",
			Message:   m[1] + " disconnected from " + m[2],
			Timestamp: entryTimestamp(entry),
		}
	}

	if m := failedRE.FindStringSubmatch(msg); m != nil {
		ip := m[2]
		if isBruteforce(ip) {
			return &core.Event{
				Source:    core.SourceSSH,
				EventType: core.EventSSHBruteforce,
				Severity:  core.SeverityWarn,
				Title:     "SSH Brute Force",
				Message:   "Repeated failed attempts for " + m[1] + " from " + ip,
				Timestamp: entryTimestamp(entry),
			}
		}
	}

	return nil
}

func isBruteforce(ip string) bool {
	bruteforceTracker.mu.Lock()
	defer bruteforceTracker.mu.Unlock()

	attempts, ok := bruteforceTracker.ips[ip]
	if !ok {
		if len(bruteforceTracker.ips) >= bruteforceMaxIPs {
			delete(bruteforceTracker.ips, bruteforceTracker.keys[0])
			bruteforceTracker.keys = bruteforceTracker.keys[1:]
		}
		attempts = nil
		bruteforceTracker.keys = append(bruteforceTracker.keys, ip)
	}

	now := time.Now()
	cutoff := now.Add(-bruteforceWindow * time.Second)

	var active []time.Time
	for _, t := range attempts {
		if t.After(cutoff) {
			active = append(active, t)
		}
	}
	active = append(active, now)
	bruteforceTracker.ips[ip] = active

	return len(active) >= bruteforceThreshold
}
